packages/task/src/browser/task-service.ts

```diff
@@ -804,24 +794,36 @@ export class TaskService implements TaskConfigurationClient {
     }
 
     protected async doRunTask(task: TaskConfiguration, option?: RunTaskOption): Promise<TaskInfo | undefined> {
+        let overridePropertiesFunction: (task: TaskConfiguration) => void = () => { };
         if (option && option.customization) {
             const taskDefinition = this.taskDefinitionRegistry.getDefinition(task);
             if (taskDefinition) { // use the customization object to override the task config
-                Object.keys(option.customization).forEach(customizedProperty => {
-                    // properties used to define the task cannot be customized
-                    if (customizedProperty !== 'type' && !taskDefinition.properties.all.some(pDefinition => pDefinition === customizedProperty)) {
-                        task[customizedProperty] = option.customization![customizedProperty];
-                    }
-                });
+                overridePropertiesFunction = tsk => {
+                    Object.keys(option.customization!).forEach(customizedProperty => {
+                        // properties used to define the task cannot be customized
+                        if (customizedProperty !== 'type' && !taskDefinition.properties.all.some(pDefinition => pDefinition === customizedProperty)) {
+                            tsk[customizedProperty] = option.customization![customizedProperty];
+                        }
+                    });
+                };
             }
         }
+        overridePropertiesFunction(task);
+        this.addRecentTasks(task);
+        try {
+            const resolver = await this.taskResolverRegistry.getTaskResolver(task.type);
+            const resolvedTask = resolver ? await resolver.resolveTask(task) : task;
+            const excutionResolver = this.taskResolverRegistry.getExecutionResolver(resolvedTask.taskType || resolvedTask.type);
+            overridePropertiesFunction(resolvedTask);
```

Comment:
Do we need to call this function twice? If we call it here again after its resolved, under what circumstances is it necessary to have called it before resolution? And why call it here, after the first resolution and after checking for an execution resolver, but before running it through the execution resolver?

Reply:
Yes, the double invocation is necessary: in particular, contributed tasks of type `npm` lose their presentation options when `TaskProvider.resolveTask()` is called. Also, VS Code seems to do the same (from reading their implementation).

Reply:
Sorry, I'm not sure I fully understand this logic with double invocation.
How it worked before your changes without double invocation?

Reply:
It did not work. That's why we were losing the presentation options ('shared' vs. 'new' terminal) with the `tsc: build` tasks. It kind of "worked" before the custom execution PR because `TaskProvider.resolve(task)` was simply never executed for any contributed task.

Reply:
anyway it looks very strange that we have to override properties:
- before `await resolver.resolveTask(task)`
- and 
- after `await resolver.resolveTask(task)`

Reply:
what's the goal to do it **before**, for example?

Reply:
"resolve()" is never supposed to be called for purely contributed tasks. So there are always customisations whenever we call resolve for a task. These customisations might influence what a plugin does during the 'resolve' call. VS code seems to do the same thing 🤷

Reply:
> resolve()" is **never** supposed to be called for purely contributed tasks. So there are **always** customisations

I don't think it's true - I can have a custom task type, custom resolver for it as well. I expect that my tasks are resolved using my resolver even if there are no customization for them.

Good example - some time ago we provided `che` tasks as `provided` or `detected` - we expected that they were resolved using `che` resolver even if there were no customization.
Now `che` tasks come as `configured` and we use `che` resolver as well. 
So  - `che` tasks as `detected` or as `configured` - we expect resolving them using our resolver - so I don't see some relations here...

Reply:
In general - it's OK for me to override properties twice if we really know what's the reason

Reply:
> I don't think it's true

It might not be true, but it should be according to the VS Code doc: https://code.visualstudio.com/api/references/vscode-api#TaskProvider

Reply:
what exactly you mean in the doc?

Reply:
Look at the documentation of the "resolveTask()" method of TaskProvider: 

> This method will not be called for tasks returned from the above provideTasks method since those tasks are always fully resolved.

Reply:
thanks!
Then
> It kind of "worked" before the custom execution PR because TaskProvider.resolve(task) was simply never executed for any contributed task type.

it was the correct behavior and now it's broken - right?

Reply:
No, the the resolve task should be called for customized task of any type that is not "shell" or "process". On the other hand, the resolution of variables should be called for any task that has an _execution_ of type `ShellExecution` or `ProcessExecution`. That behavior has never been implemented. What we have now is that we might call resolve on tasks where it is not necessary. I think that behaviour is superior to what we had before. This PR will not fix every problem in the Task system (that's why it's a roadmap item).